File: src/simulation/culture/aspect/AspectResult.java
package simulation.culture.aspect;

import kotlin.Pair;
import simulation.space.resource.MutableResourcePack;
import simulation.space.resource.tag.ResourceTag;
import simulation.space.resource.tag.labeler.ResourceLabeler;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AspectResult {
    public boolean isFinished;
    public List<Pair<ResourceLabeler, Integer>> neededResources;
    public MutableResourcePack resources;
    public ResultNode node;

    public AspectResult(
            boolean isFinished,
            List<Pair<ResourceLabeler, Integer>> neededResources,
            MutableResourcePack resources,
            ResultNode node
    ) {
        this.isFinished = isFinished;
        this.neededResources = neededResources;
        this.resources = resources;
        this.node = node;
    }

    public AspectResult(boolean isFinished, ResultNode node) {
        this(isFinished, new ArrayList<>(), new MutableResourcePack(), node);
    }

    public AspectResult(MutableResourcePack resources, ResultNode node) {
        this(true, new ArrayList<>(), resources, node);
    }

    public static class ResultNode {
        public Aspect aspect;
        public Map<ResourceTag, MutableResourcePack> resourceUsed = new HashMap<>();

        public ResultNode(Aspect aspect) {
            this.aspect = aspect;
        }
    }
}

File: src/shmp/simulation/culture/aspect/AspectResult.kt
package shmp.simulation.culture.aspect

import shmp.simulation.culture.group.centers.Group
import shmp.simulation.space.resource.container.MutableResourcePack
import shmp.simulation.space.resource.tag.ResourceTag
import shmp.simulation.space.resource.tag.labeler.ResourceLabeler
import java.util.*
import java.util.function.Consumer


data class AspectResult(
        var resources: MutableResourcePack = MutableResourcePack(),
        var node: ResultNode? = null,
        var isFinished: Boolean = true,
        var neededResources: List<Pair<ResourceLabeler, Int>> = emptyList()
) {
    fun pushNeeds(group: Group) = neededResources.forEach { (labeler, n) ->
        group.resourceCenter.addNeeded(labeler, n)
    }
}


data class ResultNode(var aspect: Aspect) {
    var resourceUsed: MutableMap<ResourceTag, MutableResourcePack> = HashMap()
}
